academy/2-cosmos-concepts/11-testing.md

```diff
@@ -0,0 +1,80 @@
+
+</HighlightBox>
+
+# Testing
+
+Testing is an integral part of software development in general. Done well, it helps teams catch bugs earlier, and thus cheaper. In the context of blockchains, which can lock large amounts of value and where [migrations](./12-migrations.md) are not simply one redeploy away from a fix, bugs can be catastrophic.
+
+The Cosmos SDK implements its own testing vision for its own modules, and it would be good if your project followed the same patterns. Following the same patterns will help everyone in the ecosystem to _speak the same language_. It will also benefit you when you open the code for bug bounty since interested bounty hunters would be _on-boarded_ faster.
+
+## Testing pyramid
+
+After having had some [reflection](https://docs.cosmos.network/main/architecture/adr-059-test-scopes.html), Cosmos divides their tests into 4 broad categories in somewhat increasing scope:
+
+* [Unit tests](https://docs.cosmos.network/main/building-modules/testing#unit-tests)
+* [Integration tests](https://docs.cosmos.network/main/building-modules/testing#integration-tests)
+* [Simulation tests](https://docs.cosmos.network/main/building-modules/testing#simulations)
+* [End-to-end tests](https://docs.cosmos.network/main/building-modules/testing#end-to-end-tests)
+
+### Unit tests
+
+Unit tests are your classic smallest tests possible and they focus on a single module at a time. If a tested call needs something from another module, like a keeper, then:
+
+* This dependency should be [mocked](https://devopedia.org/mock-testing), including mocked responses when applicable.
+* The mock should confirm that the dependency was called as expected.
+* If applicable, the test should confirm that a mocked response was handled as expected by the module under test.
+
+To be considered as well-tested, your unit tests should cover all your module's functions.
+
+### Integration tests
+
+Integration tests are one step wider in scope. It is still focused on your module, but instead of mocking the dependencies, your test provisions a fully-fledged dependency.
+
+In a well-thought-out environment, providing such a dependency should not be a concern of your module's tests. All the more so that your dependency has dependencies of its own. You want to avoid such deep correlations between modules, even their tests.
+
+This is why from version 0.47 of the Cosmos SDK, each module exposes functions to provision a minimally viable module. This way, your module only knows to call the right functions in its dependent module but no further down the line. For instance, when (integration-)testing the bank module, no staking module is provisioned as it is a side concern instead of a dependency. To facilitate these integration tests, the Cosmos team has also developed an in-house dependency injection. Note that integration tests do not create an application, but instead a chain of dependencies.
```

Comment:
```suggestion
This is why from version 0.47 of the Cosmos SDK, each module exposes functions to provision a minimally viable module. This way, your module only knows to call the right functions in its dependent module but no further down the line. For instance, when (integration-)testing the bank module, no slashing module is provisioned as it is a side concern instead of a dependency. To facilitate these integration tests, the Cosmos team has also developed an in-house dependency injection. Note that integration tests do not create an application, but instead a chain of dependencies.
```
The bank module is a wrong example, as we require staking. But this is indeed true.
Integration tests do create an application, the application just have the least module enabled.